atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/lock/RefreshCorrectnessChecker.java

```diff
@@ -0,0 +1,147 @@
+ * We assume that the events of each process in isolation are correct.
+ */
+public class RefreshCorrectnessChecker implements Checker {
+
+    private static final Logger log = LoggerFactory.getLogger(RefreshCorrectnessChecker.class);
+
+    @Override
+    public CheckerResult check(List<Event> events) {
+        Visitor visitor = new Visitor();
+        events.forEach(event -> event.accept(visitor));
+        return ImmutableCheckerResult.builder()
+                .valid(visitor.valid())
+                .errors(visitor.errors())
+                .build();
+    }
+
+    private static class Visitor implements EventVisitor {
+        private final Map<Pair<Integer, String>, InvokeEvent> pendingForProcessAndLock = new HashMap<>();
+        private final Map<Pair<Integer, String>, Event> lastHeldLock = new HashMap<>();
+        private final Map<String, TreeRangeSet<Long>> locksHeld = new HashMap<>();
+        private final ArrayList<String> allLockNames = new ArrayList<>();
+        private final List<Event> errors = new ArrayList<>();
+        private final Map<Integer, String> resourceName = new HashMap<>();
+
+        @Override
+        public void visit(InfoEvent event) {
+        }
+
+        @Override
+        public void visit(InvokeEvent event) {
+            Integer process = event.process();
+            String lockName = event.value();
+            resourceName.put(process, lockName);
+            pendingForProcessAndLock.put(new Pair(process, lockName), event);
+
+            if (!allLockNames.contains(lockName)) {
+                allLockNames.add(lockName);
+                locksHeld.put(lockName, TreeRangeSet.create());
+            }
+        }
+
+        @Override
+        public void visit(OkEvent event) {
+            Integer process = event.process();
+            String lockName = resourceName.get(process);
+            Pair processLock = new Pair(process, lockName);
+            InvokeEvent invokeEvent = pendingForProcessAndLock.get(processLock);
+
+            switch (event.function()) {
+                /**
+                 * Successful LOCK:
+                 * Remember the new value for the most recent successful lock
+                 */
+                case RequestType.LOCK:
+                    if (event.isSuccessful()) {
+                        lastHeldLock.put(processLock, event);
+                    }
+                    break;
+                /**
+                 * Successful REFRESH/UNLOCK:
+                 * Add the new interval [a, b) to the set of known locks, where
+                 * a is the OkEvent.time() of the last successful lock or the InvokeEvent.time() of refresh, and
+                 * b is the InvokeEvent.time() of the current request, if and only if b > a.
+                 * Also verify that the whole interval was free. Unlock can be treated as refresh, as the correctness
+                 * of their mutual interaction is verified by IsolatedProcessCorrectnessChecker
```

Comment:
Some thoughts on this:

- offhand, it seems a bit strange that we have `[a, b)` here but `(a, b)` in the `LockCorrectnessChecker`. There are probably reasons why this is safe (something to do with the `b > a`?) though
- would be good if we document how we pick "the last successful lock or the InvokeEvent.time() of refresh" (if I read this correctly, the first refresh adds the lock time whilst subsequent refreshes add the last refresh's invoke time)